### ISSUE TITLE
GenerateTestingPlatformEntryPoint has no effect on class library project

### ISSUE THREAD

**Issue by drieseng** [microsoft/testfx#4259]
## Describe the bug

We currently do not use **Microsoft.Testing.Platform**, and as such our test projects are defined as class library instead of executable.
Since we've upgraded to the .NET 9.0 SDK, [CA1515](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1515) is reported for our test classes. The root cause of this is that an **AutoGeneratedProgram** class is generated automatically which causes [CA1515](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1515) to believe it's dealing with an executable/application.

We tried to instruct **MSTest** (or the **Microsoft.Testing.Platform¨**) not to emit this class by setting the **GenerateTestingPlatformEntryPoint** MSBuild property to **false**, but this has no effect at all.

## Steps To Reproduce

1. Clone https://github.com/drieseng/do_not_emit_AutoGeneratedProgram.git
2. Open the **UnitTests** solution.
3. Compile the solution.
4. Open the **UnitTests** assembly in **ildasm**.

## Expected behavior

There's no **AutoGeneratedProgram** class.

## Actual behavior

There's a **AutoGeneratedProgram** class.

**Comment by Youssef1313** [comment]
That looks like VSTest. Could you try `<GenerateProgramFile>false</GenerateProgramFile>` instead?

**Comment by Evangelink** [comment]
@Youssef1313 is right, you haven't opt-in the new testing platform so `GenerateTestingPlatformEntryPoint` has no effect. I am not sure that disabling the main of VSTest would not cause issue with running tests.

@drieseng Just to confirm, are you seeing the `CA1515` on the test classes?

**Comment by Youssef1313** [comment]
It the analyzer's fault I think. This would be a duplicate of https://github.com/dotnet/roslyn-analyzers/issues/7192

**Comment by drieseng** [comment]
@Youssef1313, setting the **GenerateProgramFile** MSBuild property only makes things worse.
If I do that, the following compile error is produced:
```txt
CS5001: Program does not contain a static 'Main' method suitable for an entry point
```
I've updated my test repo accordingly.

@Evangelink, I'm indeed seeing that. I've updated my test repo to enable the analyzers. If you remove the `#pragma warning disable CA1515` from the **Class1** class, you'll see that *CA1515** is reported.

**CA1515** could indeed be updated to ignore test classes, but if the **AutoGeneratedProgram** was not emitted then **CA1515** would not have to be modified for my case.

**Comment by Youssef1313** [comment]
Yes, it's actually an analyzer's fault IMO, as I mentioned in the previous comment. Sorry for confusing you.

I don't see an issue on our side. We *could* implement a suppressor for the analyzer if really necessary, but I'm viewing this more to be an analyzer bug. My personal opinion is to go for `<NoWarn>$(NoWarn);CA1515</NoWarn>` in your test projects temporarily until discussions on what the analyzer behavior should be are settled.

**Comment by drieseng** [comment]
@Youssef1313, why don't you consider it an issue that the **AutoGeneratedProgram** class is emitted even if **GenerateTestingPlatformEntryPoint** is set to **false**?

That class looks like this:
```cs
using System;
[Microsoft.VisualStudio.TestPlatform.TestSDKAutoGeneratedCode]
class AutoGeneratedProgram {static void Main(string[] args){}}
```

Who is responsible for emitting this class, and how can I disable this?

**Comment by Youssef1313** [comment]
Because AutoGeneratedProgram is unrelated to GenerateTestingPlatformEntryPoint.

`AutoGeneratedProgram` is generated by *VSTest*, while *TestingPlatform* in GenerateTestingPlatformEntryPoint is referring to Microsoft.Testing.Platform, which is basically an alternative to VSTest. See [Microsoft.Testing.Platform overview docs](https://learn.microsoft.com/dotnet/core/testing/unit-testing-platform-intro).

The way to disable AutoGeneratedProgram in VSTest is `<GenerateProgramFile>false</GenerateProgramFile>`. But disabling will cause a lot of trouble and doesn't look like how things were designed to work.

**Comment by drieseng** [comment]
@Youssef1313, thanks for the clear feedback. One more question: how does one opt into **Microsoft.Testing.Platform**?

**Comment by Youssef1313** [comment]
See [MSTest Runner overview](https://learn.microsoft.com/en-us/dotnet/core/testing/unit-testing-mstest-runner-intro) for instructions. However, you will still be building as executable so it won't fix anything for CA1515, but it's good to migrate for other reasons, including performance.

**Comment by Youssef1313** [comment]
@drieseng I'm curious though, as the analyzer is disabled by default. Do you have something like `<AnalysisLevel>latest-All</AnalysisLevel>` that enables this rule?

**Comment by drieseng** [comment]
@Youssef1313, indeed.

**Comment by drieseng** [comment]
@Youssef1313, we actually had to revert the "migration" to the .NET 9 SDK (even when continuing to target .NET 8.0) as this causes test regressions. We still have to dig deeper and report this issue to MS.

**Comment by nohwnd** [comment]
Looking at the original project (and the first commit) I was just curious what your setup was, and how you set your projects to be a Library?

 I see you are installing the **VSTest** Microsoft.NET.Test.SDK, that package is mainly conaining .targets and has one primary goal: To make your project `<OutputType>Exe</OutputType>` so dotnet build will generate .runtime.config.json so we know which .NET Tfm you want to run and deps.json so we know what deps you have and how to locate them. 

Setting 

```
    <OutputType>Library</OutputType>
    <GenerateProgramFile>false</GenerateProgramFile>
```

Will prevent VSTest from generating the entrypoint code, but it will not prevent it from setting your library to be exe. Hence the:  CSC : error CS5001: Program does not contain a static 'Main' method suitable for an entry point

**Comment by nohwnd** [comment]
clicked the wrong button :)
